pysph/base/octree.pyx

```diff
@@ -876,18 +945,357 @@ cdef class CompressedOctree(Octree):
 
         return 1 + depth_max
     
+    @cython.cdivision(True)
+    @cython.boundscheck(False)
+    cdef int _c_build_tree_level1(self, NNPSParticleArrayWrapper pa, double* xmin, double length,
+            cOctreeNode* node, int num_threads) nogil:
+
+        cdef double* src_x_ptr = pa.x.data
+        cdef double* src_y_ptr = pa.y.data
+        cdef double* src_z_ptr = pa.z.data
+        cdef double* src_h_ptr = pa.h.data
+
+        cdef double xmin_new[8][3]
+        cdef double xmax_new[8][3]
+        cdef double hmax_children[8]
+        cdef cOctreeNode* new_node
+        cdef int i, j, k, c, tid, oct_cid, oct_id, count_thread, n, p
+        cdef u_int  q
+
+        for i from 0<=i<8:
+            hmax_children[i] = 0
+            for j from 0<=j<3:
+                xmin_new[i][j] = self.dbl_max
+                xmax_new[i][j] = -self.dbl_max
+
+
+        cdef double eps
+        n =  node.num_particles
+
+        if (n < self.leaf_max_particles):
+            for i from 0<=i<n:
+                self.pids[i] = i
+            node.is_leaf = True
+            return 1
+
+        cdef u_int* p_indices = self.pids
+        cdef vector[u_int]* child_indices = new vector[u_int](n)
+        cdef vector[vector[int]] cumulative_map = vector[vector[int]](num_threads)
+        cdef vector[vector[double]] threads_hmax = vector[vector[double]](num_threads)
+
+        cdef vector[vector[double]] threads_xmin = vector[vector[double]](num_threads)
```

Comment:
Does it make sense to put all these into a single array with 7 elements, i.e. xmin, xmax, ymin, ymax, zmin, zmax, hmax? If this makes it harder to read then leave it.